Flaky test: Test/RequestsCounter

--- FAIL: Test (2.95s)
    --- FAIL: Test/RequestsCounter (0.21s)
        --- FAIL: Test/RequestsCounter/exceeds-max-requests (0.01s)
            requests_counter_test.go:95: unexpected number of (successes, errors), expected (32, 32), encountered (33, 31)
FAIL
FAIL	google.golang.org/grpc/xds/internal/client	7.109s

From run: https://github.com/grpc/grpc-go/pull/4215/checks?check_run_id=1928775875

--- a/xds/internal/client/requests_counter.go
+++ b/xds/internal/client/requests_counter.go
@@ -56,6 +56,10 @@ func GetServiceRequestsCounter(serviceName string) *ServiceRequestsCounter {
 // StartRequest starts a request for a service, incrementing its number of
 // requests by 1. Returns an error if the max number of requests is exceeded.
 func (c *ServiceRequestsCounter) StartRequest(max uint32) error {
+	// Note that during race, the limits could be exceeded. This is allowed:
+	// "Since the implementation is eventually consistent, races between threads
+	// may allow limits to be potentially exceeded."
+	// https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking#arch-overview-circuit-break.
 	if atomic.LoadUint32(&c.numRequests) >= max {
 		return fmt.Errorf("max requests %v exceeded on service %v", max, c.ServiceName)
 	}

--- a/xds/internal/client/requests_counter_test.go
+++ b/xds/internal/client/requests_counter_test.go
@@ -91,7 +91,11 @@ func testCounter(t *testing.T, test counterTest) {
 	if test.expectedErrors == 0 && loadedError != nil {
 		t.Errorf("error starting request: %v", loadedError.(error))
 	}
-	if successes != test.expectedSuccesses || errors != test.expectedErrors {
+	// We allow the limits to be exceeded during races.
+	//
+	// But we should never over-limit, so this test fails if there are less
+	// successes than expected.
+	if successes < test.expectedSuccesses || errors > test.expectedErrors {
 		t.Errorf("unexpected number of (successes, errors), expected (%v, %v), encountered (%v, %v)", test.expectedSuccesses, test.expectedErrors, successes, errors)
 	}
 }

